File: bench/prongB/maptree.go
package main
import "fmt"
type Tree struct{ leaf bool; n int; l,r *Tree }
func leaf(n int)*Tree{ return &Tree{leaf:true,n:n} }
func node(l,r *Tree)*Tree{ return &Tree{l:l,r:r} }
func build(d int)*Tree{ if d==0 {return leaf(1)}; return node(build(d-1),build(d-1)) }
func maptree(t *Tree)*Tree{ if t.leaf {return leaf(t.n+1)}; return node(maptree(t.l),maptree(t.r)) }
func checksum(t *Tree)int{ if t.leaf {return t.n}; return checksum(t.l)+checksum(t.r) }
func main(){ t:=build(16); sum:=0; for i:=0;i<200;i++{ sum+=checksum(maptree(t)) }; fmt.Println(sum) }

File: bench/prongB/maptree.rs
enum Tree { Leaf(i64), Node(Box<Tree>, Box<Tree>) }
fn build(d:i64)->Tree{ if d==0 {Tree::Leaf(1)} else {Tree::Node(Box::new(build(d-1)),Box::new(build(d-1)))} }
fn maptree(t:&Tree)->Tree{ match t { Tree::Leaf(n)=>Tree::Leaf(n+1), Tree::Node(l,r)=>Tree::Node(Box::new(maptree(l)),Box::new(maptree(r))) } }
fn checksum(t:&Tree)->i64{ match t { Tree::Leaf(n)=>*n, Tree::Node(l,r)=>checksum(l)+checksum(r) } }
fn main(){ let t=build(16); let mut sum=0i64; for _ in 0..200 { sum+=checksum(&maptree(&t)); } println!("{}",sum); }
